### PR TITLE
ARROW-10526: [FlightRPC][C++] Client cookie middleware

### PR DESCRIPTION
**Pull request by lyndonbauto**
- Added client cookie middleware which caches and monitors expiry
- Added testing for client cookie middleware

### CODE REVIEW

**Comment by lyndonb-bq** [comment]
@lidavidm 

Hi David, when you have some time please take a look at the changes I have introduced here. As you can probably tell, this is adds a client side middleware for cookie handling in C++. I am still waiting for the CI to complete so I can make sure that it is good on all platforms, but the code structure should not change much from here unless changes are requested.

Thank you!

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:45]: Appreciate the docs, can we stick with Doxygen syntax (\brief, \param, etc) even though this won't be in the public docs?

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:54]: You may appreciate arrow::util::optional

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:55]: Please use std::string* for out parameters and const foo& for in parameters.

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:77]: Any reason why UriUnescape can't take a const std::string&?

File: cpp/src/arrow/flight/client_cookie_middleware.h
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.h:30]: pimpl pattern might be overkill here - what about a freestanding factory function that gives you a shared_ptr<ClientMiddlewareFactory>?

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:66]: Isn't this condition redundant? If no semicolon was found, then its position will always be greater than equals_pos

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:90]: Do you mean that strptime is not on Windows? There's a vendored strptime available.

Comment (lidavidm) [reply cpp/src/arrow/flight/client_cookie_middleware.cc:90]: I would much rather we not maintain an ad-hoc timestamp parser if at all possible.

**Review by lidavidm — COMMENTED** [review]
Mostly style comments here, and a comment about the timestamp parsing.

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/client_cookie_middleware.cc]: Unfortunately the vendored strptime does not support symbolics, which are required for cookie parsing. Instead of using an ad-hoc timestamp parser, I wrote a simple function to convert the symbolics to numerics, so the vendored strptime function will support it. Hope this solution is acceptable, of course I am open to other suggestions though.

**Comment by lyndonb-bq** [comment]
There appears to be a failure in the python build, I think it's caused by something unrelated to this pull request though, I noticed the same here: https://github.com/apache/arrow/runs/1456411775?check_suite_focus=true

File: cpp/src/arrow/flight/flight_test.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/flight_test.cc:1194]: This is a general comment, but there's a lot of logic here, and some of it duplicates the logic in the middleware class itself. Can we factor out the cookie parsing into an internal header and add direct tests of that logic, and perhaps reuse some of it here (e.g. splitting and parsing cookies) instead of duplicating it? That would also let us directly test the cookie parsing, instead of indirectly through these tests.

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:303]: Is there any need to discard here if we have to discard before we send cookies anyways?

File: cpp/src/arrow/flight/client_cookie_middleware.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/client_cookie_middleware.cc:191]: Can this be moved to an actual constant following the naming convention (kCookieExpiresFormat or similar)?

**Review by lidavidm — COMMENTED** [review]
My main comment here is about the tests - I'd like to see if we can refactor things so that the tests can directly test some of the cookie logic, and so that they don't duplicate said logic in the tests themselves.

Another approach might be to have glass-box tests by having an internal header with the middleware definition, so that tests can directly check the contents of the cookie cache (but so that client code still just gets an opaque middleware instance).

**Comment by lyndonb-bq** [comment]
@lidavidm That's a really good point, there is a lot of code that wasn't facing direct testing. I followed suggestion of moving the parsing logic to an internal header (the client_header_internal files seemed appropriate since this code is all related to headers). I added a bunch of tests for all the different internal parsing functions that were added.

File: cpp/src/arrow/flight/client_header_internal.h
Comment (lidavidm) [inline cpp/src/arrow/flight/client_header_internal.h:25]: nit: I don't think value_parsing is used in the header, and we're missing `#include <mutex>`, and `#include <functional>` for `binary_function`, `#include <string>`, and `#include <chrono>`

File: cpp/src/arrow/flight/flight_test.cc
Comment (lidavidm) [inline cpp/src/arrow/flight/flight_test.cc:1224]: Can we use `unordered_map` then?

File: cpp/src/arrow/flight/client_header_internal.h
Comment (lidavidm) [reply cpp/src/arrow/flight/client_header_internal.h:25]: Oh and `#include <map>` but perhaps unordered_map is better since you don't care about ordering?

**Review by lidavidm — COMMENTED** [review]
Thanks for the changes! I left a few nits, but this is good to go once fixed.

File: cpp/src/arrow/flight/flight_test.cc
Comment (lyndonb-bq) [reply cpp/src/arrow/flight/flight_test.cc:1224]: Yes, good call. Will change this.

**Review by lidavidm — APPROVED** [review]
Thanks for the changes!

**Pull request closed without merging**
